optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolver.java

```diff
@@ -149,16 +149,20 @@ public final void solve(Solution planningProblem) {
             throw new IllegalArgumentException("The planningProblem (" + planningProblem
                     + ") must not be null.");
         }
-        solverScope.setBestSolution(planningProblem);
-        outerSolvingStarted(solverScope);
-        boolean restartSolver = true;
-        while (restartSolver) {
-            solvingStarted(solverScope);
-            runPhases();
-            solvingEnded(solverScope);
-            restartSolver = checkProblemFactChanges();
+        try {
+            solverScope.setBestSolution(planningProblem);
+            outerSolvingStarted(solverScope);
+            boolean restartSolver = true;
+            while (restartSolver) {
+                solvingStarted(solverScope);
+                runPhases();
+                solvingEnded(solverScope);
+                restartSolver = checkProblemFactChanges();
+            }
+            outerSolvingEnded(solverScope);
+        } finally {
+            solving.set(false);
```

Comment:
@dimiii , I know nothing about your integration of planner, but how about catching exception in scoreFunction and returning Double.NEGATIVE_INFINITY (in right level of your score)? 

It might be completely bad, but on the other hand, you could avoid stopping planner on the first place and if using prepared heuristic for it, it shouldn't broke your solution.